File: python/test/t_RandomVectorMetropolisHastings_std.py
#! /usr/bin/env python

import openturns as ot
import openturns.testing as ott
import math

ot.RandomGenerator.SetSeed(0)

ot.TESTPREAMBLE()

# %%
# Test RandomVectorMetropolisHastings on Beta-Binomial conjugate model

# Define Beta-binomial model
a, b, lower, upper = 1.0, 1.0, 0.0, 1.0
n, p = 10, 0.5
prior = ot.RandomVector(ot.Beta(a, b, lower, upper))
model = ot.Binomial(n, p)

# %%
# Simulate data and compute analytical posterior
x = model.getSample(1)
posterior = ot.Beta(a + x[0, 0], b + n - x[0, 0], lower, upper)

# %%
# Define RVMH sampler
# Here, the random vector is defined with repsect to the prior law
# this effectively implements the IMH algorithm using the prior
# as an instrumental distribution
initialState = [p]
rvmh_sampler = ot.RandomVectorMetropolisHastings(prior, initialState, [0])
slf = ot.SymbolicFunction(["x"], [str(n), "x"])
rvmh_sampler.setLikelihood(model, x, slf)

# %%
# Generate posterior distribution sample
sampleSize = 10000
xSample = rvmh_sampler.getSample(sampleSize)

# %%
# Compare empirical to theoretical moments

ott.assert_almost_equal(
    xSample.computeMean(), posterior.getMean(), 0.0, 10.0 / math.sqrt(sampleSize)
)
ott.assert_almost_equal(
    xSample.computeStandardDeviation(),
    posterior.getStandardDeviation(),
    0.0,
    10.0 / math.sqrt(sampleSize),
)

# %%
randomVector = ot.RandomVector(ot.Normal())
initialState = [0.0]
sampler = ot.RandomVectorMetropolisHastings(randomVector, initialState)
x = sampler.getSample(10000)
mean = x.computeMean()
stddev = x.computeStandardDeviation()
ott.assert_almost_equal(mean, [0.0], 0.0, 0.03)
ott.assert_almost_equal(stddev, [1.0], 0.03, 0.0)

# with link function
slf = ot.SymbolicFunction(["x"], ["0.5", "0.1"])
sampler = ot.RandomVectorMetropolisHastings(randomVector, initialState, [0], slf)
x = sampler.getSample(10000)
mean = x.computeMean()
stddev = x.computeStandardDeviation()
ott.assert_almost_equal(mean, [0.5], 0.0, 0.003)
ott.assert_almost_equal(stddev, [0.1], 0.03, 0.0)

File: lib/test/t_RandomVectorMetropolisHastings_std.cxx
//                                               -*- C++ -*-
/**
 *  @brief The test file of class RandomVectorMetropolisHastings
 *
 *  Copyright 2005-2024 Airbus-EDF-IMACS-ONERA-Phimeca
 *
 *  This library is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU Lesser General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This library is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU Lesser General Public License for more details.
 *
 *  You should have received a copy of the GNU Lesser General Public License
 *  along with this library.  If not, see <http://www.gnu.org/licenses/>.
 *
 */
#include <iostream>
#include "openturns/OT.hxx"
#include "openturns/OTtestcode.hxx"

using namespace OT;
using namespace OT::Test;
typedef JointDistribution::DistributionCollection DistributionCollection;

int main(int, char *[])
{
  TESTPREAMBLE;
  OStream fullprint(std::cout);
  setRandomGenerator();

  try
  {
    // Test RandomVectorMetropolisHastings on Beta-Binomial conjugate model

    // Define Beta-binomial model
    Scalar a = 1.0;
    Scalar b = 1.0;
    Scalar lower = 0.0;
    Scalar upper = 1.0;
    Beta prior(a, b, lower, upper);

    UnsignedInteger n = 10;
    Scalar p = 0.5;
    Binomial model(n, p);

    // Simulate data and compute analytical posterior
    Sample X(model.getSample(1));
    Beta posterior(a + X(0, 0), b + n - X(0, 0), lower, upper);

    // Define IMH sampler
    IndependentMetropolisHastings rvmh_sampler(prior, {p}, Uniform(-1.0, 1.0), Indices(1));
    Description inVar({"x"});
    Description outVar({std::to_string(n), "x"});
    SymbolicFunction slf(inVar, outVar);
    rvmh_sampler.setLikelihood(model, X, slf);

    // Generate posterior distribution sample
    UnsignedInteger N = 10000;
    Sample Xsample(rvmh_sampler.getSample(N));

    // Compare empirical to theoretical moments
    assert_almost_equal(Xsample.computeMean(), posterior.getMean(), 0.0, 10.0 / std::sqrt(N));
    assert_almost_equal(Xsample.computeStandardDeviation(), posterior.getStandardDeviation(), 0.0, 10.0 / std::sqrt(N));

    // End Beta-Binomial conjugate model test

    // Elementary test
    Normal normal;
    const Normal & ref_normal(normal);
    RandomVector randomVector(ref_normal);
    RandomVectorMetropolisHastings sampler(randomVector, {0.0});
    Sample x(sampler.getSample(N));
    Point mean(x.computeMean());
    Point stddev(x.computeStandardDeviation());
    assert_almost_equal(mean, {0.0}, 0.0, 0.03);
    assert_almost_equal(stddev, {1.0}, 0.03, 0.0);

    // with link function

  }
  catch (TestFailed & ex)
  {
    std::cerr << ex << std::endl;
    return ExitCode::Error;
  }


  return ExitCode::Success;
}
